### PR TITLE
[FLINK-15635][table] Now EnvironmentSettings accepts the user ClassLoader

### PR DESCRIPTION
**Pull request by slinkydeveloper**
This PR solves the issue of allowing the user to provide its own classloader for UDFs and at other places where it's more appropriate than the current Thread context classloader.

The PR also contains a couple of hotfixes (mergeable separately):

* Deprecate `CatalogFunction#isGeneric`
* Remove `CodeGeneratorContext#apply` which makes the code unnecessarily more complicated
* Fix singleton config in `EnvironmentSettings`, which among various issues, it's also causing in this case to select the wrong user class loader

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogFunction.java [hunk 1]
```diff
@@ -59,6 +59,7 @@
      *
      * @return whether the function is a generic function
      */
+    @Deprecated
```
Comment (matriv) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogFunction.java:62]: Please add `@deprecated` section to the javadoc

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java [hunk 2]
```diff
@@ -621,7 +626,7 @@ public void registerTemporarySystemFunction(
     }
 
     @SuppressWarnings("unchecked")
-    private void validateAndPrepareFunction(CatalogFunction function)
+    private CatalogFunction validateAndPrepareFunction(CatalogFunction function)
```
Comment (matriv) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java:629]: Just a double check, have you checked all usages to make sure we use the returned value?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/utils/PlannerMocks.java [hunk 3]
```diff
@@ -96,6 +102,7 @@ private PlannerMocks(
                 true,
                 ExpressionResolver.resolverFor(
                         tableConfig,
+                        Thread.currentThread().getContextClassLoader(),
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/utils/PlannerMocks.java:105]: Why not use `PlannerMocks.class.getClassLoader()` here?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/StreamPhysicalJoinRuleBase.scala [hunk 4]
```diff
@@ -25,6 +25,11 @@ import org.apache.flink.table.planner.plan.nodes.{FlinkConventions, FlinkRelNode
 import org.apache.flink.table.planner.plan.utils.IntervalJoinUtil
 import org.apache.flink.table.planner.utils.ShortcutUtils.unwrapTableConfig
 
+import org.apache.flink.table.planner.plan.utils.{FlinkRelOptUtil, IntervalJoinUtil}
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/StreamPhysicalJoinRuleBase.scala:28]: leftover?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java [hunk 5]
```diff
@@ -139,6 +139,8 @@ public static String encodeObjectToString(Serializable obj) {
         }
     }
 
+    /** @see #loadClass(String, ClassLoader) */
```
Comment (matriv) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java:142]: please add `@deprecated`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java [hunk 6]
```diff
@@ -244,17 +245,23 @@ public static UserDefinedFunction instantiateFunction(Class<?> functionClass) {
     }
 
     /** Prepares a {@link UserDefinedFunction} instance for usage in the API. */
-    public static void prepareInstance(ReadableConfig config, UserDefinedFunction function) {
+    public static <T extends UserDefinedFunction> T prepareInstance(
```
Comment (matriv) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java:248]: Maybe add to the javadoc, that the caller needs to use the returned value?

**Review by matriv — COMMENTED** [review]
Looks good! Left a few comments.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java (hunk 2 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java:629]: yep, all of them now use the returned value here

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/utils/PlannerMocks.java (hunk 3 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/utils/PlannerMocks.java:105]: In test classpath, this shouldn't make any difference. Anyway, I changed it

Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/utils/PlannerMocks.java]: It's more for consistency, and avoid confusion, if someone checks the code and wonders why here is different.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java [hunk 7]
```diff
@@ -244,7 +244,11 @@ public static UserDefinedFunction instantiateFunction(Class<?> functionClass) {
         }
     }
 
-    /** Prepares a {@link UserDefinedFunction} instance for usage in the API. */
+    /**
+     * Prepares a {@link UserDefinedFunction} instance for usage in the API.
```
Comment (matriv) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java:248]: Maybe also add that it's instantiated using the `userClassLoadrer`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java [hunk 8]
```diff
@@ -139,7 +139,11 @@ public static String encodeObjectToString(Serializable obj) {
         }
     }
 
-    /** @see #loadClass(String, ClassLoader) */
+    /**
+     * @deprecated Use {@link #loadClass(String, ClassLoader)} instead, in order to explicitly
+     *     provide the correct classloader.
+     * @see #loadClass(String, ClassLoader)
```
Comment (matriv) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/EncodingUtils.java:145]: maybe, remove now this line?

**Review by matriv — APPROVED** [review]
LGTM! thx @slinkydeveloper Left a couple minor comments regarding the javadocs.

**Comment by matriv** [comment]
Thx!

**Comment by matriv** [comment]
If we don't choose to backport the functionality to 1.15, we need to backport the fix for `EnviromentSettings` with the static variables.

**Comment by lsyldliu** [comment]
@slinkydeveloper Thanks for your contribution, can you help to rebase the master and push this pr merge to master, here some other work rely on you.

**Review by lsyldliu — COMMENTED** [review]
@slinkydeveloper Thanks for your contribution, can you help to rebase the master and push this pr merge into master, some other work rely on this feature.

**Comment by matriv** [comment]
Working on rebasing and reshaping this PR

**Comment by lsyldliu** [comment]
ping @slinkydeveloper

**Comment by slinkydeveloper** [comment]
I don't think I'll be able to finish this PR, can someone take it over?

**Comment by matriv** [comment]
I'm working on it: https://github.com/matriv/flink/tree/FLINK-15635-new already.

**Comment by lsyldliu** [comment]
> 

ping @matriv Can you help to complete this work as soon as possible? [FLIP-214](https://cwiki.apache.org/confluence/display/FLINK/FLIP-214+Support+Advanced+Function+DDL) is blocked by this issue now, thanks.

**Comment by lsyldliu** [comment]
> > 
> 
> ping @matriv Can you help to complete this work as soon as possible? [FLIP-214](https://cwiki.apache.org/confluence/display/FLINK/FLIP-214+Support+Advanced+Function+DDL) is blocked by this issue now, thanks.

Hi, @matriv , If you don't have time to complete this work, I will take over it, thanks.

**Comment by lsyldliu** [comment]
Due to @matriv had not responsed for a long day, I will take over this ticket, thanks for all works.

**Comment by wuchong** [comment]
Continue the review work in #19845

**Pull request closed without merging**